### ISSUE TITLE
Terraform validate/plan succeeds but apply fails with 400 response

### ISSUE THREAD

**Issue by VRPrathapReddy** [hashicorp/terraform-provider-aws#33998]
### Description

Hi Team,

Greetings!
We recently attempted to scale down the EC2 instances for the core nodes in the cluster from `r7g.xlarge` to `r7g.large`. When the configuration change was run through `validate` and `plan` there were no issues.

Validate and Plan stages are successful because it will check for the Configuration and state files information. 
It failed at apply stage due to below reason.

API wasn't happy with the aws_emr_cluster argument and returned error 400 and a message that the r7g.large instance type was not supported.

From the API standpoint is it possible to implement this feature. Could you confirm?

Thanks and Regards,
Prathap Reddy.

### Would you like to implement a fix?

None

**Comment by VRPrathapReddy** [comment]
Error Message :

Error: running EMR Job Flow: ValidationException: Instance type 'r7g.large' is not supported.
│     status code: 400,

**Comment by breathingdust** [comment]
Hi @VRPrathapReddy, the message you are receiving is from AWS and not a limitation of Terraform. From looking at the supported instance types it does not look like `r7g.large` is supported by AWS.

- https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-supported-instance-types.html

Are you asking for increased validation at the plan/validate stage? I don't believe there is a way to programatically validate supported types.

**Comment by jar-b** [comment]
Hello 👋 

We typically avoid network calls (such as calling the [ListSupportedInstanceTypes](https://docs.aws.amazon.com/emr/latest/APIReference/API_ListSupportedInstanceTypes.html) API) within plan-time validators. However, #34438 has added a new data source wrapping this API which will allow this type of check to be accomplished with a [lifecycle precondition](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#custom-condition-checks). 

```terraform
locals {
  instance_type = "r7g.xlarge"
  release_label = "emr-6.15.0"
}

data "aws_emr_supported_instance_types" "test" {
  release_label = local.release_label
}

resource "aws_emr_cluster" "test" {
  ### other configuration ###
  release_label  = local.release_label
  master_instance_group {
    instance_type = local.instance_type
  }

  lifecycle {
    precondition {
      condition = contains(data.aws_emr_supported_instance_types.test.supported_instance_types[*].type, local.instance_type)
      error_message = "${local.instance_type} is not supported with this EMR release label!"
    }
  }
}
```

In the example above, using a supported instance type such as `r7g.xlarge` will result in a successful plan. If `instance_type` is changed to `r7g.large`, an error is returned at plan time:

```
│ Error: Resource precondition failed
│
│   on main.tf line 78, in resource "aws_emr_cluster" "test":
│   78:       condition = contains(data.aws_emr_supported_instance_types.test.supported_instance_types[*].type, local.instance_type)
│     ├────────────────
│     │ data.aws_emr_supported_instance_types.test.supported_instance_types is list of object with 515 elements
│     │ local.instance_type is "r7g.large"
│
│ r7g.large is not supported with this EMR release label!
```

Lifecycle preconditions require Terraform `>=1.2.0`, and the supported instance types data source will be available in the next AWS provider release, `v5.27.0`. I'll leave this issue open for now to allow time for feedback. If this approach provides the desired plan time validation requested in the original report, please let us know and we'll close out this issue!
